File: discord-music-bot/src/main/java/net/masterzach32/swagbot/music/LocalPlaylist.java
package net.masterzach32.swagbot.music;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import javax.sound.sampled.UnsupportedAudioFileException;

import com.mashape.unirest.http.exceptions.UnirestException;
import net.masterzach32.swagbot.App;
import net.masterzach32.swagbot.music.player.*;
import net.masterzach32.swagbot.utils.exceptions.NotStreamableException;
import net.masterzach32.swagbot.utils.exceptions.YouTubeAPIException;
import org.json.JSONObject;
import sx.blah.discord.handle.obj.IGuild;
import sx.blah.discord.handle.obj.IUser;

public class LocalPlaylist {
	
	private String name;
	private boolean locked, requiresPerms;
	private List<AudioSource> music;

	public LocalPlaylist(String name, boolean locked, boolean requiresPerms) {
		this.name = name;
		this.locked = locked;
		this.requiresPerms = requiresPerms;
		music = new ArrayList<>();
	}

	public LocalPlaylist(String name, List<YouTubeAudio> music, boolean locked, boolean requiresPerms) throws UnirestException {
		this.name = name;
		this.locked = locked;
		this.requiresPerms = requiresPerms;
		this.music = new ArrayList<>();
        for(int i = 0; i < music.size(); i++)
            this.music.add(music.get(i));
	}

	protected LocalPlaylist(JSONObject json) {
        this.name = json.getString("name");
        this.locked = json.getBoolean("locked");
        this.requiresPerms = json.getBoolean("requiresPerms");
        this.music = new ArrayList<>();
        for(int i = 0; i < json.getJSONArray("music").length(); i++) {
			if(json.getJSONArray("music").get(i) instanceof JSONObject) {
                JSONObject jsonSource = (JSONObject) json.getJSONArray("music").get(i);
                AudioSource source;
                try {
                    if (jsonSource.getString("source").equals("youtube"))
                        source = new YouTubeAudio(jsonSource.getString("url"));
                    else if (jsonSource.getString("source").equals("soundcloud"))
                        source = new SoundCloudAudio(jsonSource.getString("url"));
                    else
                        source = new AudioStream(jsonSource.getString("url"));
                    music.add(source);
                } catch (NotStreamableException | UnirestException | YouTubeAPIException e) {
                    e.printStackTrace();
                }
            } else {
                String url = (String) json.getJSONArray("music").get(i);
                AudioSource source;
                try {
                    if (url.contains("youtube"))
                        source = new YouTubeAudio(url);
                    else if (url.contains("soundcloud"))
                        source = new SoundCloudAudio(url);
                    else
                        source = new AudioStream(url);
                    music.add(source);
                } catch (NotStreamableException | UnirestException | YouTubeAPIException e) {
                    e.printStackTrace();
                }
            }
        }
    }
	
	public AudioSource add(String audio) {
		for(int i = 0; i < music.size(); i++)
		    if(music.get(i).getUrl().equals(audio))
                return music.get(i);
        AudioSource source;
        try {
            if(audio.contains("youtube"))
                source = new YouTubeAudio(audio);
            else if(audio.contains("soundcloud"))
                source = new SoundCloudAudio(audio);
            else
                source = new AudioStream(audio);
            music.add(source);
            return source;
        } catch (NotStreamableException | UnirestException | YouTubeAPIException e) {
            e.printStackTrace();
        }
        return null;
	}
	
	public void remove(String audio) {
		for(int i = 0; i < music.size(); i++)
			if(music.get(i).getUrl().equals(audio))
				music.remove(i);
	}
	
	public String getName() {
		return name;
	}
	
	public void queue(IUser user, IGuild guild) {
		Collections.shuffle(music);
		for(AudioSource s : music) {
            try {
                App.playAudioFromAudioSource(s, true, null, user, guild);
            } catch (IOException | UnsupportedAudioFileException e) {
                e.printStackTrace();
            }
		}
	}
	
	public String getInfo() {
		String str = "";
		for(int i = 0; i < music.size(); i++) {
            AudioSource source = music.get(i);
			str += "" + (i + 1) + ". **" + source.getTitle() + "**\n";
		}
		return str;
	}

	public List<AudioSource> getSources() {
        return music;
    }
	
	public int songs() {
		return music.size();
	}
	
	public boolean isLocked() {
		return locked;
	}
	
	public void toggleLocked() {
		this.locked = !locked;
	}
	
	public boolean requiresPerms() {
		return requiresPerms;
	}
	
	public void toggleRequiresPerms() {
		this.requiresPerms = !requiresPerms;
	}
}

File: discord-music-bot/src/main/java/net/masterzach32/swagbot/music/LocalPlaylist.kt
/*
    SwagBot - A Discord Music Bot
    Copyright (C) 2016  Zachary Kozar

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */
package net.masterzach32.swagbot.music

import java.io.IOException
import java.util.ArrayList
import java.util.Collections

import javax.sound.sampled.UnsupportedAudioFileException

import com.mashape.unirest.http.exceptions.UnirestException
import net.masterzach32.swagbot.App
import net.masterzach32.swagbot.guilds.GuildSettings
import net.masterzach32.swagbot.music.player.*
import net.masterzach32.swagbot.utils.exceptions.NotStreamableException
import net.masterzach32.swagbot.utils.exceptions.YouTubeAPIException
import org.json.JSONObject
import sx.blah.discord.handle.obj.IGuild
import sx.blah.discord.handle.obj.IUser

class LocalPlaylist {

    val name: String
    var isLocked: Boolean = false
        private set
    private var requiresPerms: Boolean = false
    private val music: MutableList<AudioSource>

    constructor(name: String, locked: Boolean, requiresPerms: Boolean) {
        this.name = name
        this.isLocked = locked
        this.requiresPerms = requiresPerms
        music = ArrayList<AudioSource>()
    }

    @Throws(UnirestException::class)
    constructor(name: String, music: List<AudioSource>, locked: Boolean, requiresPerms: Boolean) {
        this.name = name
        this.isLocked = locked
        this.requiresPerms = requiresPerms
        this.music = ArrayList<AudioSource>()
        music.forEach { this.music.add(it) }
    }

    constructor(json: JSONObject) {
        this.name = json.getString("name")
        this.isLocked = json.getBoolean("isLocked")
        this.requiresPerms = json.getBoolean("requiresPerms")
        this.music = ArrayList<AudioSource>()
        for (i in 0..json.getJSONArray("music").length() - 1) {
            if (json.getJSONArray("music").get(i) is JSONObject) {
                val jsonSource = json.getJSONArray("music").get(i) as JSONObject
                val source: AudioSource
                try {
                    if (jsonSource.getString("source") == "youtube")
                        source = YouTubeAudio(jsonSource.getString("url"))
                    else if (jsonSource.getString("source") == "soundcloud")
                        source = SoundCloudAudio(jsonSource.getString("url"))
                    else
                        source = AudioStream(jsonSource.getString("url"))
                    music.add(source)
                } catch (e: NotStreamableException) {
                    e.printStackTrace()
                } catch (e: UnirestException) {
                    e.printStackTrace()
                } catch (e: YouTubeAPIException) {
                    e.printStackTrace()
                }
            }
        }
    }

    fun add(audio: String): AudioSource? {
        music.indices
                .filter { music[it].url == audio }
                .forEach { return music[it] }
        val source: AudioSource
        try {
            if (audio.contains("youtube"))
                source = YouTubeAudio(audio)
            else if (audio.contains("soundcloud"))
                source = SoundCloudAudio(audio)
            else
                source = AudioStream(audio)
            music.add(source)
            return source
        } catch (e: NotStreamableException) {
            e.printStackTrace()
        } catch (e: UnirestException) {
            e.printStackTrace()
        } catch (e: YouTubeAPIException) {
            e.printStackTrace()
        }

        return null
    }

    fun remove(audio: String) {
        music.indices
                .filter { music[it].url == audio }
                .forEach { music.removeAt(it) }
    }

    fun queue(user: IUser, guild: GuildSettings) {
        Collections.shuffle(music)
        for (s in music) {
            try {
                guild.playAudioFromAudioSource(s, null, user)
            } catch (e: IOException) {
                e.printStackTrace()
            } catch (e: UnsupportedAudioFileException) {
                e.printStackTrace()
            }

        }
    }

    fun getInfo(): String {
        var str = ""
        for (i in music.indices) {
            val source = music[i]
            str += "" + (i + 1) + ". **" + source.title + "**\n"
        }
        return str
    }

    fun getSources(): MutableList<AudioSource>? {
        return music
    }

    fun songs(): Int {
        return music.size
    }

    fun toggleLocked() {
        this.isLocked = !isLocked
    }

    fun requiresPerms(): Boolean {
        return requiresPerms
    }

    fun toggleRequiresPerms() {
        this.requiresPerms = !requiresPerms
    }
}